### PR TITLE
Launch Control + Power Limit Code Review: 7/13/24

### PR DESCRIPTION
**Pull request by harleen2004**

### CODE REVIEW

File: dev/LaunchControl.c [hunk 1]
```diff
@@ -59,19 +63,23 @@ Proportional test first with other output 0, get midway with target and then tun
...
 /* Start of Launch Control */
-LaunchControl *LaunchControl_new(){
+LaunchControl *LaunchControl_new(){// this goes outside the while loop
     LaunchControl* me = (LaunchControl*)malloc(sizeof(struct _LaunchControl));
     me->slipRatio = 0;
     me->lcTorque = -1;
     me->LCReady = FALSE;
     me->LCStatus = FALSE;
-    me->pidController = (PIDController*)malloc(sizeof(struct _PIDController));
+    me->pidController = PID_new(20, 0, 0, 0.2);
```
Comment (shibcreate) [inline dev/LaunchControl.c:72]: What was the reason to do this?

File: dev/LaunchControl.c [hunk 2]
```diff
@@ -59,19 +63,23 @@ Proportional test first with other output 0, get midway with target and then tun
...
+
+ // setting PID total error @ init torque = 170 Nm?
+
     me->buttonDebug = 0;
     return me;
 }
 void slipRatioCalculation(WheelSpeeds *wss, LaunchControl *me){
-    float unfilt_speed = (WheelSpeeds_getSlowestFront(wss) / (WheelSpeeds_getFastestRear(wss))) - 1;
-    float filt_speed = unfilt_speed;
+    float4 unfilt_speed = (WheelSpeeds_getSlowestFront(wss) / (WheelSpeeds_getFastestRear(wss))) - 1;
```
Comment (shibcreate) [inline dev/LaunchControl.c:81]: Do you need floats for this calculation? What was it before

File: dev/LaunchControl.c [hunk 3]
```diff
@@ -93,16 +102,26 @@ void launchControlTorqueCalculation(LaunchControl *me, TorqueEncoder *tps, Brake
...
      if(me->LCReady == TRUE && Sensor_LCButton.sensorValue == FALSE && tps->travelPercent > .90){
         me->LCStatus = TRUE;
-        me->lcTorque = me->pidController->errorSum; // Set to the initial torque
+        me->lcTorque = me->pidController->total_error; // Set to the initial torque
         if(speedKph > 3){
-            Calctorque = calculatePIDController(me->pidController, 0.2, me->slipRatio, 0.01, mcm_Torque_max); // Set your target, current, dt
-            me->lcTorque = Calctorque; // Test PID Controller before uncommenting
-        }
+
+           PID_dtUpdate(me->pidController, 0.01);// updates the dt
```
Comment (shibcreate) [inline dev/LaunchControl.c:112]: Why are we doing this here instead of main like on SRE-7?

File: dev/LaunchControl.c [hunk 4]
```diff
@@ -93,16 +102,26 @@ void launchControlTorqueCalculation(LaunchControl *me, TorqueEncoder *tps, Brake
...
-        }
+
+           PID_dtUpdate(me->pidController, 0.01);// updates the dt 
+            //float Calctorque = calculatePIDController(me->pidController, 0.2, me->slipRatio, 0.01, mcm_Torque_max); // Set your target, current, dt
+
+           float4 PIDtorque= (float4)PID_compute(me->pidController,me->slipRatio);// we erased the saturation checks for now we just want the basic calculation
+            float4 appsTqPercent;
+            TorqueEncoder_getOutputPercent(tps, &appsTqPercent);
+            ubyte2 tor= MCM_getMaxTorqueDNm(mcm);
+            me->lcTorque = (ubyte2)(tor * appsTqPercent)+PIDtorque; // adds the ajusted value from the pid to the torqueval         }
```
Comment (shibcreate) [inline dev/LaunchControl.c:119]: What's "tor"

File: dev/LaunchControl.c [hunk 5]
```diff
@@ -111,14 +130,15 @@ void launchControlTorqueCalculation(LaunchControl *me, TorqueEncoder *tps, Brake
...
     MCM_update_LaunchControl_State(mcm, me->LCStatus);
     MCM_update_LaunchControl_TorqueLimit(mcm, me->lcTorque * 10);
-}
+     }
 bool getLaunchControlStatus(LaunchControl *me){
     return me->LCStatus;
 }
-sbyte2 getCalculatedTorque(){
-    return Calctorque;
+ubyte2 getCalculatedTorque(LaunchControl *me){
```
Comment (shibcreate) [inline dev/LaunchControl.c:137]: Why do you need to pass in a value for LC? and why change this to a ubyte2?

File: dev/bms.c [hunk 6]
```diff
@@ -447,10 +447,14 @@ ubyte2 BMS_getLowestCellVoltage_mV(BatteryManagementSystem *me)
 }
 
 ubyte4 BMS_getPackVoltage(BatteryManagementSystem *me)
+
 {
     return (me->packVoltage); 
 }
 
+ubyte4 BMS_getPackCurrent(BatteryManagementSystem *me){
+    return (me->packCurrent);
```
Comment (shibcreate) [inline dev/bms.c:456]: Did you check if packCurrent is even parsing or getting calculated?

File: dev/canManager.c [hunk 7]
```diff
@@ -737,8 +737,8 @@ void canOutput_sendDebugMessage(CanManager* me, TorqueEncoder* tps, BrakePressur
     canMessages[canMessageCount - 1].id_format = IO_CAN_STD_FRAME;
     canMessages[canMessageCount - 1].data[byteNum++] = lc->LCReady;
     canMessages[canMessageCount - 1].data[byteNum++] = lc->LCStatus;
-    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque();
-    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque() >> 8;
+    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque(lc);
```
Comment (shibcreate) [inline dev/canManager.c:740]: Is the DBC updated?

File: dev/mathFunctions.c [hunk 8]
```diff
@@ -114,3 +114,10 @@ sbyte4 swap_int32(sbyte4 val)
     val = ((val << 8) & 0xFF00FF00) | ((val >> 8) & 0xFF00FF);
     return (val << 16) | ((val >> 16) & 0xFFFF);
 }
+
+ubyte2 floorToNearestIncrement(ubyte2 value, ubyte2 increment) {
```
Comment (shibcreate) [inline dev/mathFunctions.c:118]: Need this to be sbyte or ubyte?

File: dev/motorController.c [hunk 9]
```diff
@@ -133,6 +133,14 @@ struct _MotorController
     sbyte2 LaunchControl_TorqueLimit;
     bool LCState;
 
+     sbyte2 PowerLimit_TorqueLimit;
```
Comment (shibcreate) [inline dev/motorController.c:136]: Spacing

File: dev/canManager.c [hunk 10]
```diff
@@ -737,8 +737,8 @@ void canOutput_sendDebugMessage(CanManager* me, TorqueEncoder* tps, BrakePressur
     canMessages[canMessageCount - 1].id_format = IO_CAN_STD_FRAME;
     canMessages[canMessageCount - 1].data[byteNum++] = lc->LCReady;
     canMessages[canMessageCount - 1].data[byteNum++] = lc->LCStatus;
-    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque();
-    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque() >> 8;
+    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque(lc);
+    canMessages[canMessageCount - 1].data[byteNum++] = getCalculatedTorque(lc) >> 8;
```
Comment (shibcreate) [inline dev/canManager.c:741]: LUT output debugging and PW limit debug??

File: dev/motorController.h [hunk 11]
```diff
@@ -81,6 +81,14 @@ Status MCM_getInverterStatus(MotorController* me);
 void MCM_update_LaunchControl_TorqueLimit(MotorController *me, sbyte2 lcTorqueLimit);
 void MCM_update_LaunchControl_State(MotorController *me, bool newLCState);
 
+void MCM_update_PowerLimit_TorqueLimit(MotorController *me, sbyte2 PLTorqueLimit);
+void MCM_update_LaunchControl_State(MotorController *me, bool newPLState);
```
Comment (shibcreate) [inline dev/motorController.h:85]: explain more on the reasoning for the state

File: dev/powerLimit.c [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    }
+}
+
+PowerLimit* PL_new(){
+    PowerLimit* me = (PowerLimit*)malloc(sizeof(PowerLimit));
+    me->hashtable = HashTable_new();
+    populatePLHashTable(me->hashtable); 
+    me-> PLstatus = FALSE;
+    me->pid = PID_new(1,0,0,0);// fill this in  
+     me-> powerLimittq = 0.0;
```
Comment (shibcreate) [inline dev/powerLimit.c:78]: keep spacing consistent

File: dev/powerLimit.c [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    sbyte4 rpmFloor = floorToNearestIncrement(rpm, rpmIncrement);
+    sbyte4 rpmCeiling = ceilToNearestIncrement(rpm, rpmIncrement);
+    // Retrieve torque values from the hash table for the four corners
+    ubyte4 floorFloor = get(torque_hashtable, voltageFloor, rpmFloor);
+    ubyte4 ceilingFloor = get(torque_hashtable, voltageCeiling, rpmFloor);
+    ubyte4 floorCeiling = get(torque_hashtable, voltageFloor, rpmCeiling);
+    ubyte4 ceilingCeiling = get(torque_hashtable, voltageCeiling, rpmCeiling);
+    // Error check
+    if (floorFloor == -1 || ceilingFloor == -1 || floorCeiling == -1 || ceilingCeiling == -1) {
+        return -1;
```
Comment (shibcreate) [inline dev/powerLimit.c:99]: cant return a -1 if you are doing a ubyte, need sbyte

File: dev/sensorCalculations.h [hunk 14]
```diff
@@ -100,7 +100,7 @@ void GetBrakePosition(void);
 //4 funcs, one to get each wheel's speed
 //16 bumps per rotation, 16 hz = 1 rotation per second
 #define NUM_BUMPS 16
-#define WHEEL_DIAMETER 16 //Inches
+#define WHEEL_DIAMETER 16 //Inches    /// change to 18 check it with DAQ
```
Comment (shibcreate) [inline dev/sensorCalculations.h:103]: what tire is the car running?

**Review by shibcreate — CHANGES_REQUESTED** [review]
We'll talk about it right now on zoom

File: dev/LaunchControl.c (hunk 1 above)
Comment (harleen2004) [reply dev/LaunchControl.c:72]: pass this as a param instead in the lctorquecalc()

File: dev/LaunchControl.c (hunk 4 above)
Comment (harleen2004) [reply dev/LaunchControl.c:119]: ill fix this

File: dev/LaunchControl.c (hunk 3 above)
Comment (harleen2004) [reply dev/LaunchControl.c:112]: pull this into main.

File: dev/canManager.c (hunk 7 above)
Comment (harleen2004) [reply dev/canManager.c:740]: carlie will fix

Comment (harleen2004) [reply dev/canManager.c:740]: @car1i3

File: dev/canManager.c (hunk 10 above)
Comment (harleen2004) [reply dev/canManager.c:741]: @MehulWilliams do debug pls

Comment (harleen2004) [reply dev/canManager.c:741]: @car1i3 then update dbc

File: dev/motorController.h (hunk 11 above)
Comment (harleen2004) [reply dev/motorController.h:85]: @MehulWilliams pl.c -->line 116: make variable in header for hardcoded tq limit

**Pull request merged**